errutil/error.go

```diff
@@ -40,6 +48,51 @@ func Chain(errs ...error) error {
 	}
 }
 
+// Merge reduces all given errors to a single opaque error
+// instance containing the original information from all
+// errors aggregated but in an opaque way (opposed to Chain).
+//
+// If the len(errs) == 0 it returns nil, if len(errs) == 1 it
+// returns the error itself and it will filter out nil errs.
+//
+// It is specially useful to manage lists of errors and
+// return an error only if any of the errors failed.
+func Merge(errs ...error) error {
+	return Reduce(func(err1, err2 error) error {
+		if err1 == nil && err2 == nil {
+			return nil
+		}
+		if err1 == nil {
+			return err2
+		}
+		if err2 == nil {
+			return err1
+		}
+		return errors.New(err1.Error() + ": " + err2.Error())
```

Comment:
this format usually means wrapping, right? Like `err1` would wrap `err2`?

But for what I understand that is not the intention with `Merge`, everything would be "in the same level".

Reply:
hmm interesting, I always saw formatting as orthogonal to wrapping/opaque merging. When I write errors by hand, I usually do:

```go
return fmt.Errorf("some message: %w", err)
```

If I want to wrap/expose the error through my API (errors.Is is going to work). And if I want it to be treated opaquely I do almost the same:

```go
return fmt.Errorf("some message: %v", err)
```

On the representation level they are identical (and I think they should), wrapping is only there for more detailed investigation (errors.Is/As). What I mean is that the concept of level is intrinsic to the code, errors coming from different layers are from "different levels", on the formatting we need to represent that somehow and it seems to be the same concept, like: "top call: nested err: nested2 err: etc".

I'm not entirely against thinking it as different things... just having a hard time at first, maybe you could elaborate more + propose what would feel more natural as formatting for merging 🤔

Reply:
>"top call: nested err: nested2 err: etc"
I thought this would be a "chain"?

For the "merge" action I expect just a separator. Righ now we are using colon (`:`) as a default separator, where I was thinking a comma (`,`) would make more sense. Because the errors will usually come from the same function call inside a loop with different parameters (exactly like the example).

Reply:
Ah I see....well there is the example of the loop, but actually merge is not useful only for that. You may have a chain of calls but you don't wan't to expose the internal errors through your API as is pointed out [here](https://go.dev/blog/go1.13-errors):

"When adding additional context to an error, either with fmt.Errorf or by implementing a custom type, you need to decide whether the new error should wrap the original. There is no single answer to this question; it depends on the context in which the new error is created. Wrap an error to expose it to callers. Do not wrap an error when doing so would expose implementation details."

There is a good example of when you have a database call but want to hide details on which data storage you use to callers, in that case error wrapping is not the way to go, and yet you may need to merge errors to return them (usually a single fmt.Errorf is enough, but Merge could be used on the same scenario to avoid wrapping).

I'm not against ",", I just don't think Merge should be just about a list of errors on the same level, it is an alternative to wrapping that hide details, you are chaining errors but you want the result to be opaque. But I do get your point that maybe it will be odd on the list scenario... and maybe the name Merge is not the best one for "opaque chain"....

WDYT @i4ki ?

Reply:
Ok, I got confused by the example, because you are calling the same function there, so I expect them to be almost the same error, just with different parameters. I thought that would be the main use case for `Merge`.

And I am probably missing something because I also thought you would use the `Chain` function for what you described.

Maybe you can add more (and more realistic) code examples of use cases for each one of them? (`Chain` and "opaque chain")

Reply:
Before getting into more discussion, regarding:

"And I am probably missing something because I also thought you would use the Chain function for what you described."

For what I described you can't use Chain unless you want to expose all the errors on the Chain through your API, so you definitely need both options, one where you expose errors and one where you conceil them opaquely...but now I'm not so sure if Merge would be opaque thing.

On the "more realistic examples" side of things, now that I'm thinking about it one of the main use cases for Chain is that you can't use fmt.Errorf to wrap multiple errors (you can't do fmt.Errorf("%w %w") ), for a scenario where you don't want to wrap errors you can just use fmt.Errorf("%v %v %v"), which solves for multiple levels since in general multiple levels are fixed, so just call fmt.Errorf.

Like our main problem with "level" was something like this:

```go
func parser() {
	err := lexer()
	if err != nil {
		return fmt.Errorf("parsing err %w : %w", ParserErrSentinel, err)
	}
}
```  
The whole idea is that you can check for more details on the specific nature of the error by checking the wrapped lexer error, or you can just check the more high level error which is a parser error and do the handling at that level, so we wanted to be able to navigate at different depths of errors, all wrapped (there are other ways to solve this, but we wanted something that could work with simple sentinels, no extra structs required).

But that is not possible, you can't use formatting to wrap 2 errors at once (or more), so hence it was born the Chain function, but if you want to chain that opaquely, you can just use "%v: %v"...that already works...so yeah maybe Merge is not about that at all..and it is more about lists of errors that usually happen because of some loop calling the same function.

So maybe Merge could be more about this single level thing + representing itself as a string in the same way (or similar) to a slice.

WDYT @i4ki @vitorarins ?

Reply:
Thanks for the explanation! I think that makes it much more clear. I would add the example to the `error_example_test.go`. Maybe also add the explanation as comments/docs.

For the direction of the `Merge` function let's hear more opinions from @i4ki . IMHO indeed it seems to be about a single level thing.

Reply:
From the roots of LISP: a list is a chain of cells/records. :-)

So a linked-list is a list, then talking about chain versus an opaque list of errors is indeed confusing IMO
Thanks @vitorarins for starting this discussion.

Maybe the chain of errors can be opaque or unwrapable at the user's choice.
Below are two possibilities:

1. Make the chain opaque or unwrapable by default, and a method to convert from one interface to the other:

eg. opaque by default
```
    errs = errutil.Chain(errs, err)
    errs = errutil.Chain(errs, otherErr)
    ...

    // errs implements error
    errs.Error() // reduces the chain to a single error string.
    errors.New(errs.Error()) // amounts to the current Merge()
    
    errs = errutil.Expose(errs) // convert errutil.errChain to errutil.errExposedChain
    // now errs implements the `Unwrap()` iface, so the caller can inspect the chain.
```

Example implementation:

```
type errExposedChain struct {
	errorChain
}

func Expose(errs error) error {
	err, ok := errs.(errorChain)
	if ok {
		return errExposedChain{err}
	}
	return err
}

func (e errExposedChain) Unwrap() error {
	if e.tail != nil {
		return Expose(e.tail)
	}
	return nil
}
```

Or the other way around, make it unwrapable by default and a function to make it not-unwrapable.

2. Maybe we can make it explicit that if you use `errutil.Error` then it's exposed by default. (opinionated)

eg.:

```
    errs = errutil.Chain(errors.New("not exposed"))
    errs = errutil.Chain(errors.New("also not exposed"))
    errs = errutil.Chain(ErrSomething) // exposed because it's errutil.Error
```

Example implementation (beware: not tested code):

```
func (e errorChain) Unwrap() error {
	if e.tail != nil {
		// tail is either errorChain or nil
		next := e.tail.(errorChain)
		_, ok := next.head.(Error)
		if ok {
			return e.tail
		}

		// next is not an exposed error, so advance in the chain.
		return next.Unwrap()
	}
	return nil
}
```

In both approaches, the conversion to opaque error is as trivial as calling `errs.Error()` but maybe we need to support some `Reduce()`-like function to customize how the error chain eclipses.

Not sure if this makes sense. @katcipis I already loved the current design, and no problem going ahead as it is, so feel free to ignore.

Reply:
Hmm I think the idea is interesting, will need more time to think. At first using typing to leverage the chaining, like the idea 2, is less appealing to me because it makes it hard (impossible if we hardcode a specific type) to get the Wrapping functionality when you are using your own errors types, if I have several custom errors (unrelated to errutil) and want to wrap them together I can't because they need to be of our Error type to be wrapped.  We could do the same with an interface, so the imposition/constraint is to implement an interface, but that also seems like way too much, I really wanted something that would work with anything that you are already using. We provide some helpful types, but you should not be forced to use them to get basic functionality (like wrapping).

Idea 1, which still approaches it just with functions and doesn't impose anything else on the caller seems interesting... maybe we could do the following on this PR:

* Just implement Reduce (no Merge)
* Reduce already filters nil errors, because fuck nil errors and we are not trying to do a general purpose Reduce anyway

And then we can merge that,because it is a useful way to customize how you represent a bunch of errors together as a single one (like concatenating errors with "," instead of ":" or whatever you want to do).

Reply:
I agree =) sorry for the late reply.

Reply:
sounds like a plan! :+1: